File: complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double real;
    private readonly double imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        this.real = real;
        this.imaginary = imaginary;
    }
    
    public double Real() => real;

    public double Imaginary() => imaginary;

    public ComplexNumber Mul(ComplexNumber other) =>
        new ComplexNumber(
            real * other.real - imaginary * other.imaginary,
            imaginary * other.real + real * other.imaginary
        );

    public ComplexNumber Add(ComplexNumber other) =>
        new ComplexNumber(
            real + other.real,
            imaginary + other.imaginary
        );

    public ComplexNumber Sub(ComplexNumber other) =>
        new ComplexNumber(
            real - other.real,
            imaginary - other.imaginary
        );

    public ComplexNumber Div(ComplexNumber other)
    {
        var denominator = other.real * other.real + other.imaginary * other.imaginary;

        return new ComplexNumber(
            (real * other.real + imaginary * other.imaginary) / denominator,
            (imaginary * other.real - real * real * other.imaginary) / denominator
        );
    }

    public double Abs() =>
        Math.Sqrt(real * real + imaginary * imaginary);

    public ComplexNumber Conjugate() =>
        new ComplexNumber(
            real,
            -1 * imaginary
        );

    public ComplexNumber Exp()
    {
        var factor = Math.Exp(this.real);

        return new ComplexNumber(
            Math.Cos(this.imaginary) * factor,
            Math.Sin(this.imaginary) * factor
        );
    }
}

File: fsharp/complex-numbers/ComplexNumbers.fs
module ComplexNumbers

open System

type ComplexNumber =
    { 
        real: float
        imaginary: float
    }

let private sumPows complexNumber =
    Math.Pow(complexNumber.real, 2.0) + Math.Pow(complexNumber.imaginary, 2.0)

let create real imaginary =
    {
        real = real
        imaginary = imaginary
    }

let mul z1 z2 =
    {
        real = z1.real * z2.real - z1.imaginary * z2.imaginary
        imaginary = z1.imaginary * z2.real + z1.real * z2.imaginary
    }

let add z1 z2 =
    {
        real = z1.real + z2.real
        imaginary = z1.imaginary + z2.imaginary
    }

let sub z1 z2 =
    {
        real = z1.real - z2.real
        imaginary = z1.imaginary - z2.imaginary
    }

let div z1 z2 = 
    {
        real = (z1.real * z2.real + z1.imaginary * z2.imaginary) / (z2 |> sumPows)
        imaginary = (z1.imaginary * z2.real - z1.real * z2.imaginary) / (z2 |> sumPows)
    }

let abs z =
    Math.Sqrt(z |> sumPows)

let conjugate z =
    { z with imaginary = -z.imaginary }

let real z =
    z.real

let imaginary z =
    z.imaginary

let exp z =
    {
        real = Math.Cos(z.imaginary) * Math.Exp(z.real)
        imaginary = Math.Sin(z.imaginary) * Math.Exp(z.real)
    }